api/src/main/java/org/apache/iceberg/expressions/InclusiveMetricsEvaluator.java

```diff
@@ -372,6 +372,38 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
       return ROWS_MIGHT_MATCH;
     }
 
+    @Override
+    public <T> Boolean notStartsWith(BoundReference<T> ref, Literal<T> lit) {
```

Comment:
if notStartsWith can be negated from startsWith, why not replace the whole method body with "return !startsWith(ref, lit);", instead of replicating so much code?

Reply:
notStartsWith can't necessarily be stood in for not(startsWith) depending on the situation.

Here's a quote from @rdblue on the original ticket.

```
Yes, I think we will need a NotStartsWith to handle this.

We can't keep the Not after rewriting because some evaluation doesn't support negation. For example, when lower and upper bounds are used to evaluate whether to read a file, the result may be true to indicate that the value may be contained in the file. That should not be negated because it does not confirm that the value is definitely in the file: if looking for files that "do not contain X", we can't use "not(file might contain X)".
```

However, in situations where we believe it can be, I would agree the reducing unneeded code is a good way to go. But given that we're dealing with ROWS_MIGHT_MATCH, my understanding is that we're not necessarily able to negate that as all rows don't necessarily match. So rows in a file could very well be `ROWS_MIGHT_MATCH` for startsWith as well as for notStartsWith.

Reply:
Here's the ticket for your reference: https://github.com/apache/iceberg/issues/1952

Reply:
`ResidualEvaluator` is a case where we could simply return `!startsWith(ref, lit)`. That's actually how I defined it (using the same ternary expression by swapping the return values on true or false). But given that function body for `ResidualEvaluator` is only one line, it's arguable whether or not that's worth it as it could bring confusion about which cases can we use `!startsWith` as a stand in.

Reply:
One way to reduce the code bloat / repetition would be to define a function `private boolean startsWith(ByteBuffer value, ByteBuffer prefix, Comparator<ByteBuffer> cmp)` (or possibly returning int to allow it to be used inside of the `startsWith` code), but that would ultimately just remove the repetition of the three lines of code that truncate the upper / lower bound to match the byte length of the prefix and then compare. The functions would still need to be different from each other.

I considered doing that in this PR, but I felt that changing the implementation of `startsWith` at the same time would possibly be a bit too many changes for one PR (as this PR is already large). I'd be happy to clean it up to make the code a bit cleaner, especially in a follow up PR, but I don't think that we can use `!startsWith(ref, lit)` for really any of these cases other than `ResidualEvaluator` unfortunately 🙁 .

Reply:
The other place where `!startsWith` would be correct is in the current implementation of `BoundLiteralPredicate`, though again that's a simple one line (and I've also marked down some concerns I have with the implementation as it currently is written). In the case of `BoundLiteralPredicate` though, regardless of the final implementation of `startsWith`, since we're not dealing with `file might contain x` but just a single value comparison, `!startsWith` would be correct. But again, it's only one line and we don't save much in terms of complexity.

Outside of that, it seems that substituting `not(startsWith)` wouldn't work. As mentioned, most of the time we've asssumed that any `NOT` nodes have been removed from the AST of the query via `RewriteNot`.

Reply:
I'm going to mark this as closed to make it easier for people to review this. As discused in a comment elsewhere with @rdblue and @RussellSpitzer, places where substitution of `notStartsWith`'s definition as `!startsWith` have been updated. The remaining ones require the usage of a `notStartsWith` specific definition due to the above contraints in regards to what is being tested for (i.e. we're testing in those cases for `file might match`).

If anybody has any concerns feel free to comment here or elsewhere 👍